File: nuido_flow_data/flows/data/data_group_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
# 
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
# 
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.

import logging

_logger = logging.getLogger(__name__)
import ast
import pandas as pd

from odoo import fields as fds, models
from odoo.tools import date_utils as dtu, DEFAULT_SERVER_DATE_FORMAT

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode

class DataGroupNode(BaseNode):
    def process(self, params):
        super().process(params)

        domain = []
        if ("domain" in params):
            domain = params["domain"]

        data = self._query_data(domain)

        return {
                self.definition["key"]: data
            }


    def _query_data(self, additional_domain):
        model_name = self.definition["model"]
        model = self.env[model_name]

        node_def = self.definition

        field_infos = node_def["fields"]

        domain = ast.literal_eval(self.definition["domain"])
        domain = domain + additional_domain

        dynamic_date_field = self.definition["dynamic_date_field"]
        dynamic_date_interval = self.definition["dynamic_date_interval"]

        if dynamic_date_field != "" and dynamic_date_interval != "":
            now = fds.Datetime.today()
            dynamic_domain = [(dynamic_date_field, ">=" , dtu.start_of(now, dynamic_date_interval)),
                            (dynamic_date_field, "<=", dtu.end_of(now, dynamic_date_interval))]

            domain = domain + dynamic_domain

        datetime_granularity = node_def["datetime_granularity"]

        group_field_name = node_def["group_field"]
        group_field_meta = self.env["ir.model.fields"]._get(model_name, group_field_name)

        aggr_funcs = []
        orders = []
        value_field_metas = []

        group_by = group_field_name
        if group_field_meta.ttype in ["date", "datetime"]:
            group_by = f"{group_by}:{datetime_granularity}"

        for field_info in field_infos:
            value_field_name = field_info["technical"]
            value_field_meta = self.env["ir.model.fields"]._get(model_name, value_field_name)
            value_field_metas.append(value_field_meta)

            aggr_func = f"{value_field_name}:{node_def["aggregate_function"]}"
            aggr_funcs.append(aggr_func)

            orders.append(f"{aggr_func} asc")

        order = f"{group_by} desc"
        order = order + "," + ",".join(orders)

        aggs = model.sudo()._read_group(
            domain=domain,
            groupby=[group_by],
            aggregates=aggr_funcs,
            order=order
        )

        group_field_display_name = group_field_meta.field_description
        value_field_display_names = [o.field_description for o in value_field_metas]
        df = pd.DataFrame(aggs,
                        columns=[
                            group_field_display_name,
                            *value_field_display_names
                        ]
                    )

        df = df.map(lambda o: o.name if isinstance(o, models.Model) else o)
        res = df

        output_type = self.definition["output_type"]

        if output_type == "array":
            res = df.to_dict("records")
        elif output_type == "html":
            res = res.to_html()

        return res

File: nuido_flow_data/static/src/components/data/data_group_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
// 
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
// 
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { useState } from "@odoo/owl";
import { Node } from "@nuido/components/node";
import { ModelSelectorEx } from "@nuido_flow/components/ui/model_selector_ex";
import { ModelFieldSelectorEx } from "@nuido_flow/components/ui/model_field_selector_ex";
import { ModelFieldTags } from "@nuido_flow/components/ui/model_field_tags";
import { DomainDialogInput } from "@nuido_flow/components/ui/domain_dialog_input";
export class DataGroupNode extends Node {
    setup() {
        super.setup();
        this.state = useState({
            model: this.props.node.model,
            modelDescription: this.props.node.model_description
        });
    }
    get modelSelectorId() {
        return `input-${this.props.node.id}-model-selector`;
    }
    onModelSelected(model) {
        const { label, technical } = model;
        this.props.node.fields = [];
        this.state.model = technical;
        this.state.modelDescription = label;
        this.props.node.model = technical;
        this.props.node.model_description = label;
    }
    onFieldDeleted(fieldName) {
        const idx = this.props.node.fields.findIndex(o => o.technical === fieldName);
        if (idx > -1) {
            this.props.node.fields.splice(idx, 1);
        }
        this.refreshEdges();
    }
    onFieldAdded(fieldInfo) {
        this.props.node.fields.push(fieldInfo);
        this.refreshEdges();
    }
    onDomainChanged(domain) {
        this.props.node.domain = domain;
    }
    dynamicDateFieldFilter(value) {
        return value.searchable && (["date", "datetime"].find(o => o === value.type));
    }
    onDynamicDateFieldUpdate(path, info) {
        this.props.node.dynamic_date_field = info.fieldDef ? info.fieldDef.name : "";
    }
    onDynamicDateIntervalChanged(ev) {
        this.props.node.dynamic_date_interval = ev.target.value;
    }
    onOutputTypeChanged(ev) {
        this.props.node.output_type = ev.target.value;
    }
    onGroupFieldUpdate(path, info) {
        this.props.node.group_field = info.fieldDef ? info.fieldDef.name : "";
    }
    onAggregateFunctionChanged(ev) {
        this.props.node.aggregate_function = ev.target.value;
    }
    onDatetimeGranularityChanged(ev) {
        this.props.node.datetime_granularity = ev.target.value;
    }
    get aggregateFunctionInputId() {
        return `input-${this.props.node.id}-aggregate-function`;
    }
    get datetimeGranularityInputId() {
        return `input-${this.props.node.id}-datetime-granularity`;
    }
}
DataGroupNode.template = "nuido_flow_data.data-group-node";
DataGroupNode.components = {
    ...Node.components,
    ModelSelectorEx,
    DomainDialogInput,
    ModelFieldSelectorEx,
    ModelFieldTags
};
